pkg/controller/storagecluster/reconcile.go

```diff
@@ -287,6 +312,76 @@ func (r *ReconcileStorageCluster) reconcileNodeTopologyMap(sc *ocsv1.StorageClus
 	return nil
 }
 
+// ensureNodeRacks iterates through the list of storage nodes and ensures
+// all nodes have a rack topology label.
+func (r *ReconcileStorageCluster) ensureNodeRacks(nodes *corev1.NodeList, minRacks int, nodeRacks, topologyMap *ocsv1.NodeTopologyMap, reqLogger logr.Logger) error {
+
+	for _, node := range nodes.Items {
+		hasRack := false
+
+		for nodeName := range nodeRacks.Labels {
+			if nodeName == node.Name {
+				hasRack = true
+				break
+			}
+		}
+
+		if !hasRack {
+			rack := determinePlacementRack(node.Name, minRacks, nodeRacks)
+			reqLogger.Info("Labeling node with rack label", "Node", node.Name, "Label", defaults.RackTopologyKey, "Value", rack)
+			node.Labels[defaults.RackTopologyKey] = rack
+			nodeRacks.Add(node.Name, rack)
+			if !topologyMap.Contains(defaults.RackTopologyKey, rack) {
+				reqLogger.Info("Adding topology label from node", "Node", node.Name, "Label", defaults.RackTopologyKey, "Value", rack)
+				topologyMap.Add(defaults.RackTopologyKey, rack)
+			}
+
+			reqLogger.Info("Updating node", "Node:", node.Name)
+			err := r.client.Update(context.TODO(), &node)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// determinePlacementRack sorts the list of known racks in alphabetical order,
+// counts the number of Nodes in each rack, then returns the first rack with
+// the fewest number of Nodes. If there are fewer than three racks, define new
+// racks so that there are at least three.
+func determinePlacementRack(nodeName string, minRacks int, nodeRacks *ocsv1.NodeTopologyMap) string {
```

Comment:
this is super tricky when we are on something like AWS.

because AWS's EBS volumes are pinned to the AZ they exist in, the underlying OSD storage is pinned to the AZ regardless of the concept of "racks"

So, this means we could build a rack of nodes, but the osd on that rack couldn't actually fail over (isn't portable) if those nodes are in reality on different AWS AZs. This wouldn't be obvious at all until a failure event occurred. 

Basically, each rack can only be made of nodes within the same zone when the zone label exists.

So take a configuration like 6 nodes spread evenly across 2 zones.

- node1 at zone-east-a
- node2 at zone-east-a
- node3 at zone-east-a
- node4 at zone-east-b
- node5 at zone-east-b
- node6 at zone-east-b

The 3 racks would have to be something like this.

- rack1 - node1,node2, node3
- rack2 - node4, node5
- rack3 - node6

even though we have a node count divisible by 3, the distribution when 2 AZs are used won't be optimal.

Reply:
...fair point. I think I can do something about it, though.

Reply:
Oh, definitely, a "rack" of nodes needs to contained in a single AZ.